docs/recap/bitvm2_fast_recap.tex

```diff
@@ -0,0 +1,229 @@
+    transaction has passed verification, and the operator can spend the
+    output, thereby confirming the statement \(f(x) = y\).
+  \item \textbf{DisproveScript} --- one of the challengers has found a
+    discrepancy in the intermediate states \(z_i\), \(z_{i-1}\) and the
+    sub-program \(f_i\). In other words, they have proven that
+    \(f_i(z_{i-1}) \neq z_i\), and thus, they can spend the output.
+\end{enumerate}
+
+\subsection{\texttt{DispoveScript}}\label{sec:dispove-script}
+
+\texttt{DispoveScript} is part of the MAST tree in a Taproot address,
+which allows challengers to claim the transaction amount for states
+\(z_i\), \(z_{i-1}\), and sub-program \(f_i\), is called
+\(DisproveScript_i\) in BitVM2:
+
+\begin{verbatim}
+// push z_i, z_{i-1} onto the stack
+{ z_i     }
+{ z_(i-1) }
+// compute f(z_{i-1})
+{ f_i     }
+// ensure that f_i(z_{i-1}) != z_i
+OP_EQUAL
+OP_NOT
+\end{verbatim}
+
+In fact, this script does not need a \texttt{script\_sig}, as with the
+correct \(z_i\) and \(z_{i-1}\), it will always execute
+successfully. Therefore, to restrict spending capability, a Lamport
+signature verification is added to the script.
+
+\subsection{Lamport signature}\label{sec:lamport-signature}
+
+Unlike other digital signature algorithms, the Lamport signature uses
+a pair of random secret and public keys \((sk_{\mathcal{M}},
+pk_{\mathcal{M}})\) that can sign and verify only any message from the
+space \(\mathcal{M} = {\{0, 1\}}^{l}\) of \(l\) -bit messages.
+
+However, once the signature \(c_{m}\) is formed, where \(m\) is the message
+being signed, \((sk_{\mathcal{M}}, pk_{\mathcal{M}})\) become tied to \(m\),
+because any other signature with these keys will compromise the keys
+themselves. Thus, for the message \(m\), the keys \((sk_{\mathcal{M}},
+pk_{\mathcal{M}})\) are one-time use. From now on, we will denote them as
+\((sk_{m}, pk_m)\).
+
+Thus, by taking the intermediate states \(z_i\) as the message for the
+Lamport signature, we form one-time key pairs: \((sk_{z_0,
+}pk_{z_0})\), \((sk_{z_1, }pk_{z_1})\), \(\ldots\), \((sk_{z_{n-1},
+}pk_{z_{n-1}})\) and signatures \(c_{z_1}\), \(\ldots\),
+\(c_{z_{n-1}}\), respectivly. Then, to spend the output from the
+\texttt{Assert} transaction using the \(DispoveScript_i\) script,
+challenger is required adding the corresponding states \(z_i\),
+\(z_{i-1}\), and signatures \(c_{z_{i}}\), \(c_{z_{i-1}}\) to the
+stack in the \texttt{script\_sig}, making the \texttt{script\_sig} of
+the transaction input like this:
+
+\begin{verbatim}
+// push z_i, z_{i-1} onto the stack
+{ z_i         }
+{ z_(i-1)     }
+// push c_{z_i}, c_{z_{i-1}} onto the stack
+{ c_(z_i)     }
+{ c_(z_(i-1)) }
+\end{verbatim}
+
+By adding the signatures verification code and public keys to the
+\texttt{DisproveScript}s, they become:
+
+\begin{verbatim}
+// Push the keys corresponding to states z_i, z_{i-1}
+{ pk_(z_i)      }
+{ pk_(z_(i-1))  }
+// Verify the Lamport signature
+{ CheckLampComm }
```

Comment:
I assume lamport verification takes `pk` and `c` as an input, so the script most likely becomes
```
c_(z_i)
pk_(z_i)
OP_CHECKLAMPCOMM
c_(z_(i-1))
pk_(z_(i-1))
OP_CHECKLAMPCOMM
```

Reply:
But it's the `script_pubkey` part, not the combined with `script_sig` one. I specifically made them sequential with the script above (`script_sig`), so it's easier to visually concatenate them.

I assume here, that `CheckLampComm` verifies signatures for both keys and signatures, as it's not that simple to show in opcodes.  Especially in case, when state is not a single stack element. For me it's easier to write in that way 😅

P.S. : Do you think that `OP_CHECKLAMPCOMM` is better, even if Bitcoin doesn't have an opcode for that?

Reply:
Let us revisit this notation as we progress with actually verifying Lamport signatures on-chain.

P.S. I usually prefer calling key operations (even those not natively supported by Bitcoin) by using `OP_SOMETHING`. But it is just my personal taste, having nothing in common with the standard notation.

Reply:
I makes sense as `{}` looks the same as pushing something, so if we want to distinguish them that great. But I'm personally not sure how to use show the execution of the subprogram `OP_F_{n}` ?